packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts

```diff
@@ -52,4 +78,53 @@ export class FunctionUrlOrigin extends cloudfront.OriginBase {
       originKeepaliveTimeout: this.props.keepaliveTimeout?.toSeconds(),
     };
   }
-}
\ No newline at end of file
+}
+
+/**
+ * An Origin for a Lambda Function URL with OAC.
+ */
+export class FunctionUrlOriginWithOAC extends cloudfront.OriginBase {
+  private originAccessControl?: cloudfront.IOriginAccessControl;
+  private functionUrl: lambda.IFunctionUrl;
+
+  constructor(lambdaFunctionUrl: lambda.IFunctionUrl, props: FunctionUrlOriginWithOACProps = {}) {
+    const domainName = cdk.Fn.select(2, cdk.Fn.split('/', lambdaFunctionUrl.url));
+    super(domainName, props);
+    this.functionUrl = lambdaFunctionUrl;
+    this.originAccessControl = props.originAccessControl;
+
+    validateSecondsInRangeOrUndefined('readTimeout', 1, 180, props.readTimeout);
+    validateSecondsInRangeOrUndefined('keepaliveTimeout', 1, 180, props.keepaliveTimeout);
+  }
+
+  protected renderCustomOriginConfig(): cloudfront.CfnDistribution.CustomOriginConfigProperty | undefined {
+    return {
+      originSslProtocols: [cloudfront.OriginSslPolicy.TLS_V1_2],
+      originProtocolPolicy: cloudfront.OriginProtocolPolicy.HTTPS_ONLY,
+    };
+  }
+
+  public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
+    const originBindConfig = super.bind(scope, options);
+    const distributionId = options.distributionId;
+
+    if (!this.originAccessControl) {
+      this.originAccessControl = new cloudfront.FunctionUrlOriginAccessControl(scope, 'FunctionUrlOriginAccessControl');
+    }
+
+    new lambda.CfnPermission(scope, `InvokeFromApiFor${options.originId}`, {
```

Comment:
Can we move this into a private function? It would make it clearer and easier to understand why a new permission is getting created

Reply:
Also is it possible to use any existing grant or addPermission methods here instead? And how do we handle permissions for imported lambda functions?

Reply:
It seems like there might still be a potential bug with addPermission, so I’m using cfnPermission instead. It’s not too complex, and since the cross-stack test is passing, I don't have any concerns. Here’s a reference (apologies, it's in Japanese!) https://dev.classmethod.jp/articles/cdk-over-21-lambda-create-error/.

Reply:
fixed.
https://github.com/aws/aws-cdk/pull/31339/commits/2ad91705267267ed03c9ab85a1b54d96c7e6bf80